quimb/evo.py

```diff
@@ -422,6 +436,35 @@ def int_step_callback(t, y, H):
                 step_callback(t, pt, H)
 
         self._step_callback = step_callback
+
+        # if there is an integration stopping condition, set up the function
+        #   to use this
+
+        if int_stop is not None:
+            int_stop_no_tryexcept = int_stop
+
+            # try just passing time and state to a single callback function
+            # if it fails, pass time, state and hamiltonian
+            def int_stop(t, y, H):
```

Comment:
Since this is in several places, worth writing a decorator that tries a function with three args and then two? If you made it a class decorator, you could then cache whether the function is a 2 or 3 arg function. Something like this (untried):
```python
class Try3Then2Args:                       
                                           
    def __init__(self, fn):                
        self.fn = fn                       
        self.num_args = None               
                                           
    def first_call(self, p, t, H):         
        try:                               
            res = self.fn(p, t, H)         
            self.num_args = 3              
        except TypeError:                  
            res = self.fn(p, t)            
            self.num_args = 2              
            # also catch other errors etc  
        return res                         
                                           
    def __call__(self, p, t, H):           
        if self.num_args is None:          
            return self.first_call(p, t, H)
        elif self.num_args == 2:           
            return self.fn(p, t)           
        elif self.num_args == 3:           
            return self.fn(p, t, h)               
```
Might help with all the nested definitions.

Reply:
I've done something very similar to this. Currently a `Try2Then3Args`  is defined outside of the `Evolution` class even though it's only used in there. I thought it looks neater that way.